src/hdf5/read_dpr_gpm.py

```diff
@@ -0,0 +1,162 @@
+            dname = dname.replace('method', 'method1')
+            xr_dims[dname] = unique_dims[dname]
+        xr_data[k] = xr.DataArray(data[k], dims=xr_dims)
+
+    epoch_time = np.zeros(xr_data.nscan.size)
+    for iscan in xr_data.nscan.values:
+        time1 = datetime.datetime(xr_data['FS/ScanTime/Year'].values[iscan],
+                                  xr_data['FS/ScanTime/Month'].values[iscan],
+                                  xr_data['FS/ScanTime/DayOfMonth'].values[iscan],
+                                  xr_data['FS/ScanTime/Hour'].values[iscan],
+                                  xr_data['FS/ScanTime/Minute'].values[iscan],
+                                  xr_data['FS/ScanTime/Second'].values[iscan],
+                                  xr_data['FS/ScanTime/MilliSecond'].values[iscan])
+        epoch_time[iscan] = round((time1 - epoch).total_seconds())
+
+    # build output dictionary
+    dprdata = xr.Dataset()
+    dprdata['lat'] = xr_data['FS/Latitude']       # "nscan,nray=nfov";
+    dprdata['lon'] = xr_data['FS/Longitude']      # "nscan,nray=nfov";
+    dprdata['height'] = xr_data['FS/PRE/height']  # "nscan,nray=nfov,nbin=nelev";
+    dprdata['epoch_time'] = xr_data['FS/navigation/scLat'].copy()
+    dprdata['epoch_time'].values = np.squeeze(epoch_time)
+    dprdata['obs_measured'] = xr_data['FS/PRE/zFactorMeasured']  # "nscan,nray=nfov,nbin=nelev,nfreq=nchan"
+    dprdata['obs'] = xr_data['FS/SLV/zFactorFinal']  # "nscan,nray=nfov,nbin=nelev,nfreq=nchan"
+    dprdata['obs'].values[dprdata['obs'].values < -100] = np.nan
+    dprdata['zenith_angle'] = xr_data['FS/PRE/localZenithAngle']  # nscan,nray=nfov,nfreq=nchan
+    dprdata['azimuth_angle'] = dprdata['zenith_angle'].copy() * 0
+    dprdata['centerFreq'] = xr.DataArray(np.array([13.6, 35.5]), dims={"nfreq": 2})  # GHz
+    dprdata['centerWN'] = xr.DataArray(np.array([0.453, 1.183]), dims={"nfreq": 2})  # 1/cm
+
+    # rename dimensions
+    dprdata = dprdata.rename({'nbin': 'elevation', 'nfreq': 'channel', 'nray': 'fov', 'nscan': 'scanline'})
+    dprdata['fov1'] = dprdata.fov.copy()
+
+    return dprdata
+
+
+def read_dpr_gpm(dpr_fnames,
+                 dpr_dir=[],
+                 dt_start=datetime.datetime(1900, 1, 1),
+                 dt_end=datetime.datetime(2100, 1, 1)):
+
+    if len(dpr_fnames) == 0:
+        # files are partitioned into almost two hours period
+        dt_start1 = dt_start - datetime.timedelta(hours=2)
+        dt_end1 = dt_end + datetime.timedelta(hours=2)
+        dt_range = pd.date_range(dt_start1, dt_end1, freq="1H")
+        dpr_fnames = []
+        for idt in dt_range:
+            # 2A.GPM.DPR.V9-20211125.20170907-S075404-E092639.020033.V07A.HDF5
+            file_pattern = '*.%s*.HDF5' % idt.strftime('%Y%m%d-S%H')
+            for dir, _, _ in os.walk(dpr_dir):
+                dpr_fnames.extend(glob(os.path.join(dir, file_pattern)))
+
+    if not dpr_fnames:
+        return []
+
+    # limit data so that dt_start <= date <= dt_end
```

Comment:
we generally do not limit times here as this is easily accomplished within the UFO

it does mean files may have some overlapping data in the IODA files if say even 6-hourly files are created